socketioxide/src/handler/extract.rs

```diff
@@ -244,53 +241,42 @@ impl<A: Adapter> FromMessage<A> for Bin {
     type Error = Infallible;
     fn from_message(
         _: Arc<Socket<A>>,
-        _: serde_json::Value,
-        bin: Vec<Vec<u8>>,
+        mut v: PayloadValue,
         _: Option<i64>,
     ) -> Result<Self, Infallible> {
-        Ok(Bin(bin))
+        Ok(Bin(v.extract_binary_payloads()))
```

Comment:
I've left in the `Bin` extractor, and preserved its behavior.  Is it worth keeping?

- There is an example in a doc comment where the `Bin` extractor is used, and then the binary payloads are emitted along with three different messages, separate from the data in two of them.  Maybe that is a use-case that is useful to preserve?
- A downside of keeping it is that `Data` would be much more efficient if it implemented `FromMessage` and not `FromMessageParts`, since with the latter we have to clone the entire thing.
- If we do keep it, maybe `Bin` should instead implement `FromMessageParts` and clone the binary payloads rather than extracting them from the `PayloadValue`.  Then `Data` could implement `FromMessage`, if we assume that the most common use case is that people will use `Data` and not `Bin`; that would make the common case more efficient, removing the clone of the data.

 I'm leaning toward keeping it and implementing that last point above.

Reply:
One wrinkle: if I convert `Data` to `FromMessage`, I'm thinking about `TryData`, and maybe leaving it as `FromMessageParts`.  Maybe someone wants to do a `TryData`, but if it fails, still get the data in another form in order to do something else with it.  Like:

```rust
socket.on("test", |TryData(data): TryData<MyData>, Data(value): Data<PayloadValue>| async {
    if let Ok(data) = data {
        println!("Got data as expected: {:?}", data);
    } else {
        println!("Data deser failed, payload received was: {:?}", value);
    }
})
```

Or even implement a handler that can have the data deserialize into multiple possible known types:

```rust
|TryData(my_data): TryData<MyData>, TryData(other_data): TryData<OtherData>, Data(last_chance): Data<LastChanceData>|
```

Thoughts?

Reply:
![image](https://github.com/Totodore/socketioxide/assets/119153472/33b17a8f-cac2-4af3-9f7d-40c3c2ffc824)
```
fn on_handle(socket: SocketRef, ack: AckSender, Bin(bin): Bin, Data(data): Data<PayloadValue>) {
    info!("Received message: {:?}", data);
    ack.send(data.clone()).ok();
    socket.emit("message-back", data.clone()).ok();
}
```
The picture is the printed data, and the data received in on_handle is decoded from binary to string.

Reply:
Yep, I know things don't really work yet ;)  I've already fixed this locally (seems `String`'s default serde impl handles bytes, sigh), but am working on some other fixes as well before I push it.

Reply:
谢谢,辛苦了

Reply:
> I've left in the `Bin` extractor, and preserved its behavior. Is it worth keeping?
> 
> * There is an example in a doc comment where the `Bin` extractor is used, and then the binary payloads are emitted along with three different messages, separate from the data in two of them.  Maybe that is a use-case that is useful to preserve?
> * A downside of keeping it is that `Data` would be much more efficient if it implemented `FromMessage` and not `FromMessageParts`, since with the latter we have to clone the entire thing.
> * If we do keep it, maybe `Bin` should instead implement `FromMessageParts` and clone the binary payloads rather than extracting them from the `PayloadValue`.  Then `Data` could implement `FromMessage`, if we assume that the most common use case is that people will use `Data` and not `Bin`; that would make the common case more efficient, removing the clone of the data.
> 
> I'm leaning toward keeping it and implementing that last point above.

I think the best would be to remove the `Bin` extractor and the `bin()` operator and have `Data` and `TryData` implement `FromMessage` at least for `Data`. I still don't know what we should do for `TryData` maybe we could start by only implement `FromMessage` and then later having a custom implementation of `FromMessageParts` that clones the data. But we still need to think about this.

Reply:
Ok, removed `Bin`, and made `Data` implement `FromMessage`.  I left `TryData` as `FromMessageParts` for now, but we can revisit that.